src/internal/m/m.go

```diff
@@ -0,0 +1,441 @@
+// per-IP, per-request), but this system has no such limitation.  Cardinality has no impact on write
+// performance.  Analysis code can make the decision on which fields to discard to decrease the
+// cardinality for long-term storage, if desired.
+//
+// Aggregating each operation on a metric over time recovers the value of the metric at a particular
+// time.  Any sort of smartness or validation comes from the reader, not from this writing code.  If
+// you want to treat a certain metric name as a string gauge, integer counter, and sampler of proto
+// messages, that is OK.  The analysis code that processes the logs will need to be ready for that,
+// or at least ready to ignore values it doesn't think are valid.
+//
+// To emit metrics, simply call these public functions in this package:
+//
+//	Set(ctx, "metric_name", value) // Set the current value of the metric to value.
+//	Inc(ctx, "metric_name", delta) // Increment the current value of the metric by delta.
+//	Sample(ctx, "metric_name", sample) // Add sample to the set of samples captured by the metric.
+//
+// It is safe to write to the same metric from different goroutines concurrently.
+//
+// Some minimal aggregation can be done in-process.  This is a compromise to reduce the "noise" in
+// the logs.  If you were uploading a 1GB file, it would make sense to increment the byte count
+// metric 1 billion times, as each byte of the file is passed to the TCP stack.  This, however,
+// would be very noisy and make the logs difficult to read.  So, we offer aggregates to flush the
+// value of gauges (Set) and counters (Inc) periodically, grouping many value-changing operations
+// into a single log message.
+//
+// An aggregator can be registered on a context (with pctx.Child), causing all future writes to that
+// metric on that context to be aggregated.  (The code that writes the metric need not be aware of
+// the aggregated nature; the above public API automatically does the right thing.)
+//
+// Aggregated metrics are emitted to the logs based on a time interval set at registration time.  If
+// a write occurs, and the metric hasn't been logged for that interval, then a log line will be
+// produced showing the current value of the metric.  If unflushed data exists when the context is
+// Done, a log line is also emitted.  From an analysis standpoint, nothing changes; the emitted
+// aggregated metrics are indistinguishable from immediately emitted metrics.
+//
+// Aggregated metrics can only be one type of metric with one type of value; if you create a counter
+// named tx_bytes and then call Set on it, no aggregation will be done on that data.  (Inc calls
+// continue to be aggregated; it doesn't "break" the aggregation to accidentally call the wrong
+// value-changing function.)  Similarly, the type of the value is set at registration time; calls
+// writing values of a different type will not be aggregated, but again, do not break any ongoing
+// aggregation.
+package m
```

Comment:
I appreciate the value of this  being short because it is used often, but I think it goes against the guidelines of not choosing package names, which would be good, short variable names.